pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+// TelnetConn is the struct representing the telnet connection
+type TelnetConn struct {
+	// conn is the underlying connection
+	conn              net.Conn
+	readCh            chan []byte
+	writeCh           chan []byte
```

Comment:
Yes! I re-thought the design of the components. There were definitely some components that does nothing more than relay data from component A to component B. With the latest commit, the FSM reads from the connection directly which eliminates the readLoop go routine and the readCh field in the TelnetConn struct. Same thing for the the read-side part of the connectionLoop. The writeCh (and the writeLoop goroutine) still exists though because there can be multiple writers to the connection. (to clarify, there is a single writer to the dataReadWriter.dataBuffer but there are multiple writers to the underlying TCP connection)